### PR TITLE
Add new traces: `scattermap`, `choroplethmap` and `densitymap` which use `maplibre` to render maps

### PR DESCRIPTION
**Pull request by archmoj**
Supersedes #7015.
Thanks to @birkskyum's PR, I kept mapbox traces and added new traces i.e. `scattermap`, `choroplethmap` and `densitymap` to the API. 
These traces use `map`, `map2`, etc. on the `layout` and do **not** require a `config` parameter as they use [maplibre-gl](https://github.com/maplibre/maplibre-gl-js).

@plotly/plotly_js

### CODE REVIEW

**Comment by archmoj** [comment]
Symbols don't render using `scattermap`.
```js
Plotly.newPlot(gd, {
    data: [{
        type: "scattermap",
        mode: "markers+text+lines",
        lon: [-75, -80, -50], 
        lat: [45, 20, -20],
        marker: {
            size:20, 
            symbol: ["bus", "harbor", "airport"]
        }
    }]
});
```  
I asked @birkskyum for help: https://github.com/plotly/plotly.js/pull/7015#issuecomment-2252880795

File: src/plots/map/constants.js [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+                visibility: 'visible'
+            }
+        },
+    ],
+
+};
+
+var stylesMap = {
+    basic: cartoVoyager,
+    streets: cartoVoyager,
```
Comment (emilykl) [inline src/plots/map/constants.js:131]: @gvwilson @archmoj I feel we should remove the `streets` and `outdoor` style options for the `Map` trace types since we no longer have a distinct map type for them. However I'm not sure whether it's better to do it now or later. Maybe we could add a deprecation warning? Or simply remove them, and document that users will need to switch to another style when swapping `Mapbox` for `Map`.

File: src/components/modebar/manage.js [hunk 2]
```diff
@@ -110,6 +110,7 @@ function getButtonGroups(gd) {
     var hasGL2D = fullLayout._has('gl2d');
     var hasTernary = fullLayout._has('ternary');
     var hasMapbox = fullLayout._has('mapbox');
+    var hasMap = fullLayout._has('map');
```
Comment (emilykl) [inline src/components/modebar/manage.js:113]: @archmoj I noticed that in some places you have `hasMap = layoutOut._has('mapbox') || layoutOut._has('map');` and in others `hasMap = fullLayout._has('map');` -- is there a reason? Any easy way to make these consistent?

File: src/plots/map/constants.js [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+    outdoors: cartoVoyager,
+    light: cartoPositronNoLabels,
+    dark: cartoDarkmatterNoLabels,
```
Comment (emilykl) [inline src/plots/map/constants.js:134]: shouldn't this be simply `cartoDarkmatter`? I'm noticing that the original 
mapbox_0 baseline has labels, but the new [map_0](https://github.com/plotly/plotly.js/blob/929be3ff8df19b5d17512924da2d50ce9ba4fc93/test/image/baselines/map_0.png#diff-e384cac154f6ae8fc9e94c7c47c213e8b2a55b4387a1d6188aa28af4991725f2) baseline has no labels.

File: src/plots/map/constants.js [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+        },
+    ],
+
+};
+
+var stylesMap = {
+    basic: cartoVoyager,
+    streets: cartoVoyager,
+    outdoors: cartoVoyager,
+    light: cartoPositronNoLabels,
```
Comment (emilykl) [inline src/plots/map/constants.js:133]: ditto for this one, shouldn't it be `cartoPositron`? Compare 
[mapbox_bubbles-text](https://github.com/plotly/plotly.js/blob/master/test/image/baselines/mapbox_bubbles-text.png)
 baseline (has labels) with [map_bubbles-text](https://github.com/plotly/plotly.js/blob/929be3ff8df19b5d17512924da2d50ce9ba4fc93/test/image/baselines/map_bubbles-text.png) baseline (no labels).
)

File: src/plots/map/constants.js (hunk 1 above)
Comment (gvwilson) [reply src/plots/map/constants.js:131]: :+1: on a deprecation warning

File: test/image/mocks/map_custom-style.json
Comment (emilykl) [inline test/image/mocks/map_custom-style.json:1]: @archmoj This is not testing the same thing as [mapbox_custom-style.json](https://github.com/plotly/plotly.js/blob/master/test/image/mocks/mapbox_custom-style.json) -- needs to pass a custom style URL to `style`!

File: test/image/baselines/map_layers.png
Comment (emilykl) [inline test/image/baselines/map_layers.png:1]: @archmoj This one looks incorrect to me -- [mapbox_layers.png](https://github.com/plotly/plotly.js/blob/master/test/image/baselines/mapbox_layers.png) renders a red polygon outline on the left map.

File: test/image/baselines/pattern_bars.png
Comment (emilykl) [inline test/image/baselines/pattern_bars.png:1]: @archmoj Any idea what changed in these next 7 baselines and why? I can't see any visual difference.

File: test/image/compare_pixels_test.js [hunk 5]
```diff
@@ -107,16 +105,13 @@ for(var i = 0; i < allMockList.length; i++) {
...
-        'mapbox_osm-style'
-    ].indexOf(mockName) !== -1) {
-        continue;
-    }
+    if(blacklist.indexOf(mockName) !== -1) continue;
 
     var isMapbox = mockName.substr(0, 7) === 'mapbox_';
     var isOtherFlaky = [
         // list flaky mocks other than mapbox:
+        'map_density0-legend',
```
Comment (emilykl) [inline test/image/compare_pixels_test.js:113]: @archmoj why are these here rather than in the blacklist above?

File: test/jasmine/tests/modebar_test.js [hunk 6]
```diff
@@ -1406,6 +1445,50 @@ describe('ModeBar', function() {
             });
         });
 
+        describe('map handlers', function() {
```
Comment (emilykl) [inline test/jasmine/tests/modebar_test.js:1448]: @archmoj All appearances of `mapbox` in this test should be changed to `map`, no?

Or is this specifically testing the interaction between `mapbox` buttons and `map` traces?

File: test/image/baselines/pattern_bars.png
Comment (archmoj) [reply test/image/baselines/pattern_bars.png:1]: This is just a noise because we added new baselines without starting their names with `zz`.

**Comment by LiamConnors** [comment]
@birkskyum how does using the Stadia tiles work with this. If you use `Densitymap` and "stamen-watercolor", you need to specify a token?

**Comment by birkskyum** [comment]
@LiamConnors , yes, that the key would be in the url like this:

`style: "https://tiles.stadiamaps.com/styles/stamen_watercolor.json?api_key=YOUR-API-KEY"`

Docs here:
https://docs.stadiamaps.com/map-styles/stamen-watercolor/

**Comment by LiamConnors** [comment]
> @LiamConnors , yes, the key to the tile provider would be in the url like this:
> 
> `style: "https://tiles.stadiamaps.com/styles/stamen_watercolor.json?api_key=YOUR-API-KEY"`
> 
> Docs here: https://docs.stadiamaps.com/map-styles/stamen-watercolor/

Thanks @birkskyum! For someone using Plotly.js, does that mean you wouldn't ever pass just the style name like this for stamen maps (example from the current version of Plotly.js)

```
var data = [{type: 'densitymapbox', lon: [10, 20, 30], lat: [15, 25, 35], z: [1, 3, 2]}];

var layout = {width: 600, height: 400, mapbox: {style: 'stamen-watercolor'}};

var config = {mapboxAccessToken: "<token>"};

Plotly.newPlot('myDiv', data, layout, config);
```
And the example would instead look like this?


```
var data = [{type: 'densitymapbox', lon: [10, 20, 30], lat: [15, 25, 35], z: [1, 3, 2]}];

var layout = {width: 600, height: 400, mapbox: {style: 'https://tiles.stadiamaps.com/styles/stamen_watercolor.json?api_key=YOUR-API-KEY'}};

Plotly.newPlot('myDiv', data, layout);
```

**Comment by birkskyum** [comment]
@LiamConnors  yes, you got the syntax exactly right. For paid tiles/styles, the key will always go in the url like that. For free resources, no key is needed. I don't have a full overview of who hosts tiles for various stamen maps, and what prices they have and if there are free ones.

**Comment by emilykl** [comment]
@LiamConnors @archmoj Seems we should _either_ remove the `stamen-` style options entirely, or add a `mapToken` config option which gets appended to the end of the Stadia URLs the way the `mapboxAccessToken` is currently.

**Comment by LiamConnors** [comment]
> @LiamConnors @archmoj Seems we should _either_ remove the `stamen-` style options entirely, or add a `mapToken` config option which gets appended to the end of the Stadia URLs the way the `mapboxAccessToken` is currently.

The token would be nice as it would make it easier to update existing code.

**Comment by ndrezn** [comment]
@archmoj I don't see (but maybe I'm missing it) a deprecation notice for `mapbox` type traces in this PR. Will you add that here or open a separate PR for the deprecation notice?

File: test/image/baselines/map_layers.png
Comment (birkskyum) [reply test/image/baselines/map_layers.png:1]: just to clarify, that polygon came from one of mapbox's terrain endpoints - i didn't have a replacement data source. Also, maplibre has a more flexible way to drawing contour lines, which is the recommendation, where you can set the breakpoints, labels etc. - https://maplibre.org/maplibre-gl-js/docs/examples/contour-lines/

File: src/components/modebar/manage.js (hunk 2 above)
Comment (archmoj) [reply src/components/modebar/manage.js:113]: Good call. Addressed in [5fc241b](https://github.com/plotly/plotly.js/pull/7060/commits/5fc241bf09eb3200588737963c39e0a2e74787cc).

File: src/plots/map/constants.js (hunk 3 above)
Comment (archmoj) [reply src/plots/map/constants.js:134]: @birkskyum What do you think of this?

File: test/image/mocks/map_custom-style.json
Comment (archmoj) [reply test/image/mocks/map_custom-style.json:1]: This one seems outdated. @emilykl Could we resolve it?

File: test/image/baselines/map_layers.png
Comment (archmoj) [reply test/image/baselines/map_layers.png:1]: @emilykl Good catch!
@birkskyum Thanks for the note. Would you please open a PR to fix rendering of this polygon?

File: test/image/compare_pixels_test.js (hunk 5 above)
Comment (archmoj) [reply test/image/compare_pixels_test.js:113]: In the blacklist we didn't produce the baselines in the previous step. 
https://github.com/plotly/plotly.js/blob/c8f57028c0cfff778b472f5449a520670472933b/test/image/make_baseline.py#L97

So we basically skip them in the comparison test. 
For flaky ones we create the baselines and compare them with some tolerance.

File: test/image/baselines/map_layers.png
Comment (archmoj) [reply test/image/baselines/map_layers.png:1]: @birkskyum Is there a way to modify the new maplibre mock to display the polygon?

File: test/image/mocks/map_custom-style.json
Comment (emilykl) [reply test/image/mocks/map_custom-style.json:1]: @archmoj It doesn't look like the issue has been resolved -- the test still seems to be passing built-in style names to the `style` parameter, rather than passing a URL to a custom style.

File: src/plots/map/layout_attributes.js [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+        valType: 'any',
+        values: constants.styleValuesMap,
+        dflt: constants.styleValueDflt,
+        description: [
+            'Defines the map layers that are rendered by default below the trace layers defined in `data`,',
+            'which are themselves by default rendered below the layers defined in `layout.map.layers`.',
+            '',
+            'These layers can be defined either explicitly as a Map Style object which can contain multiple',
+            'layer definitions that load data from any public or private Tile Map Service (TMS or XYZ) or Web Map Service (WMS)',
+            'or implicitly by using one of the built-in style objects which use WMSes',
```
Comment (LiamConnors) [inline src/plots/map/layout_attributes.js:40]: @birkskyum I was wondering about this: "map layers that are rendered by default below the trace layers defined in `data`"

When testing the following example out on the Python side:

```

import pandas as pd
import plotly.graph_objects as go

quakes = pd.read_csv('https://raw.githubusercontent.com/plotly/datasets/master/earthquakes-23k.csv')


fig = go.Figure(go.Densitymap(lat=quakes.Latitude, lon=quakes.Longitude, z=quakes.Magnitude, radius=10))

fig.update_layout(
    map={
        "style": "carto-darkmatter",
        "center": {"lon": 180},
    },
    margin={"r": 0, "t": 0, "l": 0, "b": 0}
)

fig.show()
```
I see this, where it looks like the text is above the data. Is that expected.

image

The previous implementation looked like this

image

Comment (birkskyum) [reply src/plots/map/layout_attributes.js:40]: When adding a layer to mapbox/maplibre, it's possible to specify which layer in the style it should sit below - this is thus dependent on which style is used.  In this case, with the carto dark matter style, that setting should likely be 'waterway_label' to give the expected outcome. The primary tool used to examine the individual layers of a style.json, and make new styles, is [Maputnik](https://maplibre.org/maputnik/?layer=2138096197%7E0#1.05/0/0).§

**Comment by archmoj** [comment]
@ndrezn @emilykl `attributions` added in #4069 (See https://github.com/plotly/plotly.js/pull/4069/files#diff-084161b3ee075801aa5bf2cc0ea0b232ea5a770a76a98785a09b132f858c36f2) was required in respect to mapbox.
But using `maplibre` in new traces I think the desirable default behavior should be not have any attribution over the map subplot. What's your opinion on it?

File: src/plots/map/layout_attributes.js (hunk 7 above)
Comment (archmoj) [reply src/plots/map/layout_attributes.js:40]: I personally like this new look; but we should possibly try to make the maplibre version look similar to the mapbox one. @birkskyum Do you know what may change in the mock or code to make that happen?

File: test/jasmine/tests/modebar_test.js (hunk 6 above)
Comment (archmoj) [reply test/jasmine/tests/modebar_test.js:1448]: @emilykl 
This seems to be fixed.
If so please resolve it.

File: test/jasmine/tests/modebar_test.js [hunk 8]
```diff
@@ -1406,6 +1445,50 @@ describe('ModeBar', function() {
...
+
+                    return Plotly.relayout(gd, {
+                        'map.zoom': 10,
+                        'map.center.lon': 30
+                    });
+                })
+                .then(function() {
+                    _assert(30, 10, 10);
+
+                    var button = selectButton(gd._fullLayout._modeBar, 'resetViewMapbox');
```
Comment (emilykl) [inline test/jasmine/tests/modebar_test.js:1481]: @archmoj Is this supposed to say `Mapbox` here or should it be `Map`?

Comment (archmoj) [reply test/jasmine/tests/modebar_test.js]: Thanks. Now I see this other test :pray:
Addressed by [1dd8c0e](https://github.com/plotly/plotly.js/pull/7060/commits/1dd8c0e58c6ac87bef3d10198606b374bf951bbf).

File: test/jasmine/tests/modebar_test.js (hunk 6 above)
Comment (archmoj) [reply test/jasmine/tests/modebar_test.js:1448]: Resolved by [1dd8c0e](https://github.com/plotly/plotly.js/pull/7060/commits/1dd8c0e58c6ac87bef3d10198606b374bf951bbf).

File: src/plots/map/constants.js (hunk 3 above)
Comment (birkskyum) [reply src/plots/map/constants.js:134]: If the original had labels, then cartoDarkmatter is the right option.

File: test/image/baselines/map_layers.png
Comment (birkskyum) [reply test/image/baselines/map_layers.png:1]: If we can extract the geojson with the vertices of the polygon, then it's possible to draw it with plotly/maplibre.

File: test/image/mocks/map_custom-style.json
Comment (archmoj) [reply test/image/mocks/map_custom-style.json:1]: https://demotiles.maplibre.org/style.json

Comment (archmoj) [reply test/image/mocks/map_custom-style.json:1]: Which looks like: https://demotiles.maplibre.org/

File: src/plots/map/constants.js (hunk 3 above)
Comment (archmoj) [reply src/plots/map/constants.js]: Good catch!
Fixed in [6a46b43](https://github.com/plotly/plotly.js/pull/7060/commits/6a46b43874c25ee68541beb518ac0706f9fffd45).

**Review by emilykl — APPROVED** [review]
Looks great. 👍 🌟 💫

**Comment by birkskyum** [comment]
> Not sure if this was resolved but I agree that the desirable default is no attribution. @birkskyum do you have any thoughts on this?

The icon in the corner is entirely optional, so if it's preferable to the product to leave it out, then that seems like the right thing to do. Like with other BSD/MIT/Apache licensed software, the only ask is really to [retain/reproduce the copyright notice](https://github.com/maplibre/maplibre-gl-js/blob/main/LICENSE.txt) somewhere, but apart from that the only concern is what's best for the end-user.  In the case of plotly.js, where each graph typically is rather small, and there are often multiple graphs, then the icon takes up more space / makes more noise than i.e. a full screen app like https://maps.bing.com (they also opted not to show the icon, but it would have taken up relatively less space).

**Comment by archmoj** [comment]
@birkskyum @alexcjohnson 
FYI - This PR is approved by @emilykl and other team members to release a RC.
Remaining improvements e.g. hiding attributions by default, removing constant stamen styles keys that require API key, deprecation notices should be addressed in upcoming PRs after the release of this RC.
